posts/inside-rust/2022-04-20-libs-aspirations.md

```diff
@@ -0,0 +1,209 @@
+
+Where exactly we draw the line between things that should and shouldn't go in the standard library
+can be tricky to define and is somewhat flexible, but there are a few categories we're most interested in.
+
+**Ergonomics**
+
+A lot of additions to the standard library are very small ones that increase ergonomics.
+Quite often, these are things that were already possible in some way, just not in an ergonomic way.
+Some recent examples are:
+
+- `abs_diff()`
+- `Path::is_symlink`
+- `iter::from_fn`
+- `NonZero*::saturating_add`
+
+While we always have to consider the trade-off for niche features to the already large interface
+on some types and traits, additions like these continue to happen regularly.
+
+**Standardizing some bigger features the ecosystem needs**
+
+As Rust grows into new territories, there is more and more a need
+for certain features to be included in the standard library.
+This is especially true for things where a consistent, standard, interface is
+important.
+Some of the bigger examples are:
+
+- Async traits and functions
+- Allocators and fallible allocation
+- Error and panic handling
+- Portable SIMD
+
+**Reducing and improving unsafe code**
+
+By providing the right low level APIs and abstractions, we can greatly minimize
+the amount of complex unsafe code that users need to write. Tools like
+`MaybeUninit` guide users to correct unsafe code that's easy to follow and
+prove correct. Even better, some APIs can entirely remove the need for unsafe
+code in many situations.
+This includes situations where users tend to reach for `unsafe` for performance reasons.
+
+- `std::arch`
+- `std::simd`
+- Scoped threads
+- More atomic primitives
+- '`Iterator`' with static length for arrays
+- Improving `MaybeUninit` and related methods
+- Extending `NonNull` and pointer methods
+- A more complete interface to `OsString`, `Path`, and `CString`
+- Documentation for `Pin` and other 'unsafe' types
+- File descriptors and handles (`OwnedFd`, etc.)
+
+### Improving implementations of things within the standard library
+
+Historically, the implementation details of the standard library got less attention
+than its public API.
+Recently, however, we're seeing more and more contributions towards improving the implementation
+of various parts of the standard library.
+
+These are some parts that we're especially interested in seeing improvements in:
+
+- `core::fmt` and the implementation of `format_args!()` and `fmt::Arguments`
+- [Synchronization primitives like `Mutex`, `RwLock`, and `Condvar`](https://github.com/rust-lang/rust/issues/93740)
+- Cleanups in platform-specific code in `std::sys`
+- Avoiding allocations wherever possible, [such as when calling functions in `std::fs`](https://github.com/rust-lang/rust/pull/93668)
+- Making widely used types [such as `std::io::Error`](https://github.com/rust-lang/rust/pull/87869) more light-weight
+- Cleaning up all unnecessary `SeqCst` memory ordering
```

Comment:
The generic thread parker uses it too much, it would be nice if it had the same ordering guarantees as the futex parker, but getting it all the way there is tricky.